Refactor FXIOS-6368 [v115] Refactor Notification service to not use tab queue (#14305)
mozilla-mobile/firefox-ios: Docs/notification-service.md

> # Notification Service and Push

Author: I don't have permission to edit the wiki, and I wanted to make sure this document gets a quick review so I uploaded here.

I'll remove it once we add it to the wiki and before landing this PR

Reviewer: Added https://github.com/mozilla-mobile/firefox-ios/wiki/Notification-Service-and-Push

The [Notification Service](https://developer.apple.com/documentation/usernotifications/modifying_content_in_newly_delivered_notifications) extension is a an app extension that allows Firefox iOS to intercept encrypted payloads from [Autopush](https://autopush.readthedocs.io/en/latest/), decrypting them, and possibly modifying them before displaying them to the user.

This document will cover how the Notification Service is setup, its interaction with both the Autopush and Firefox Accounts system to deliver send tabs.

The document will also cover how it's possible to extend the notification service and push to support other, non Firefox Accounts related push notifications.

## Requirements

> To understand why the Notification Service and push work the way they do today, it's important to understand some of the constraits around how Firefox Accounts delivers push notifications.

Reviewer 2: ```suggestion
To understand why the Notification Service and push work the way they do today, it's important to understand some of the constraints around how Firefox Accounts delivers push notifications.
```

Firefox Accounts leverages the [Autopush](https://autopush.readthedocs.io/en/latest/) infrastructure to deliver push notifications. Because it uses Autopush, **Firefox Accounts delivers encrypted payloads to the user, that need to be decrypted before being displayed.**

## How a Push Notification is delivered to the User

The above requirement thus requires that the application:

> 1. **Register with Autopush**, and recieve a unique URL to pass to Firefox Accounts so Firefox Accounts can send push notifications to the appllications.

Reviewer 2: ```suggestion
1. **Register with Autopush**, and receive a unique URL to pass to Firefox Accounts so Firefox Accounts can send push notifications to the applications.
```

2. **Generate a public/private encryption key pair** to be used to encrypt the communication from Firefox Account to the application.
3. **Deliver the public encryption key to Firefox Accounts.**
4. **Persist the private encryption key.**

> 5. When Firefox Accounts sends a push notification to Autopush, it will encrypt the message using the public key it recieved from the application.

Reviewer 2: ```suggestion
5. When Firefox Accounts sends a push notification to Autopush, it will encrypt the message using the public key it received from the application.
```

8. The application would then **display the notification**

At the bottom of this document is a sequence diagram visualizing the above interactions.

Step 7 above is what is done by the Notification Service, because Apple requires that if the application would like to modify the content of a push notification before displaying to the User, the application must use a Notification Service app extention.

There is some additional complexity in Step 7 this document will cover next, particularly that **Send Tabs aren't delived directly through the push notification**, instead, the push notification carries the send tab's index - so the notification service must query Firefox Accounts to get the actual tab's URL and title.

## How a Send Tab is delivered to the User, then opened by the application
Send tabs are a form of [Device Commands](https://mozilla.github.io/ecosystem-platform/relying-parties/how-tos/device-registration#device-commands). At the time of this document, they are the only device command used by Firefox.

> Device commands don't carry their payload, instead, they carry an index to where the application can requrest their payload. For example, for send tab, the push notification (after it's decrypted by push) can look like the following:

Reviewer 2: ```suggestion
Device commands don't carry their payload, instead, they carry an index to where the application can request their payload. For example, for send tab, the push notification (after it's decrypted by push) can look like the following:
```

```json
{
  version: 1,
  command: "fxaccounts:command_received",
  data: {
    "command": "https://identity.mozilla.com/cmd/open-uri",
    "sender": "0a4abb5e6f2e378f3aadda7f97482e99",
    "url": "https://api.accounts.firefox.com/v1/device/commands?index=42&limit=1"
  }
}
```

> The application must then query the `url` to recieve the content of the actual tab (there is a second layer of encryption there, but that's out of scope for this document).

Reviewer 2: ```suggestion
The application must then query the `url` to receive the content of the actual tab (there is a second layer of encryption there, but that's out of scope for this document).
```
